File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/domain/RefundEligibilityService.java
package com.example.accountservice.payment.domain;

/**
 * Domain Service — a pure class with no framework annotations (not registered as a Spring bean; the
 * Application layer creates it directly with {@code new} when needed. See root
 * docs/architecture/domain-service.md).
 *
 * <p>The judgment "the original payment must be in COMPLETED status, and the refund amount cannot
 * exceed the payment amount" cannot be made by {@link Payment} alone, nor by {@link Refund} alone.
 * {@code Payment} does not know about refund attempts against itself (a refund exists only as a
 * {@code Refund} Aggregate), and {@code Refund} does not know the original payment's amount/status
 * (it only references it via {@code paymentId}). Making this judgment requires loading both
 * Aggregates and comparing them together, so this coordination logic cannot be placed as a method
 * on either Aggregate (doing so would require it to accept the entire other Aggregate as a
 * parameter, breaking the boundary) — this is exactly where a Domain Service belongs.
 *
 * <p>{@code mlFraudRiskScore} is a plain value already computed upstream by {@code
 * RefundFraudRiskScorer} (a Technical Service trained on the requester's actual refund/payment
 * *history* — refund count, rejection count, amount ratio, minutes since payment; structured facts
 * the requester cannot simply type something different to fake). This method never calls it and
 * doesn't know an ML model produced the value; it only weighs the fraud-risk signal alongside its
 * other checks and still owns the actual judgment.
 */
public class RefundEligibilityService {

    // A signal produced upstream by RefundFraudRiskScorer (a Technical Service trained on
    // refund/payment history, see infrastructure/RefundFraudRiskScorerNativeImpl /
    // RefundFraudRiskScorerHttpImpl) — this Domain Service never calls it and doesn't know an ML
    // model produced it. It only receives the already-computed score as one more plain input
    // alongside Payment/Refund, and applies its own fixed threshold. The model supplies a signal;
    // this method still owns the actual approve/reject judgment.
    private static final double ML_FRAUD_RISK_REJECTION_THRESHOLD = 0.8;

    public RefundDecision evaluate(Payment payment, Refund refund, double mlFraudRiskScore) {
        if (payment.getStatus() != PaymentStatus.COMPLETED) {
            return RefundDecision.rejected(
                    PaymentException.ErrorCode.REFUND_REQUIRES_COMPLETED_PAYMENT,
                    "A refund can only be requested for a completed payment.");
        }
        if (refund.getAmount() > payment.getAmount()) {
            return RefundDecision.rejected(
                    PaymentException.ErrorCode.REFUND_AMOUNT_EXCEEDS_PAYMENT,
                    "The refund amount cannot exceed the payment amount.");
        }
        if (mlFraudRiskScore >= ML_FRAUD_RISK_REJECTION_THRESHOLD) {
            return RefundDecision.rejected(
                    PaymentException.ErrorCode.REFUND_PATTERN_FLAGGED_HIGH_RISK,
                    "This refund pattern was flagged as high risk by the fraud-risk model and"
                            + " requires manual review.");
        }
        return RefundDecision.approve();
    }
}

File: implementations/nestjs/examples/src/payment/domain/refund-eligibility-service.ts
import { Payment } from '@/payment/domain/payment'
import { PaymentStatus } from '@/payment/payment-enum'
import { PaymentErrorMessage } from '@/payment/payment-error-message'
import { Refund } from '@/payment/domain/refund'

export interface RefundDecision {
  readonly approved: boolean
  readonly reason?: string
}

// The fraud-risk score is produced upstream by RefundFraudRiskScorer (a Technical Service
// trained on the requester's own refund/payment history — see
// refund-fraud-risk-scorer-native-impl.ts / refund-fraud-risk-scorer-http-impl.ts). This Domain
// Service never calls it and doesn't know how it was computed; it only receives the
// already-computed score as a plain number alongside Payment/Refund, and applies its own fixed
// threshold. The Technical Service supplies a signal; this method still owns the actual
// approve/reject judgment.
const ML_FRAUD_RISK_REJECTION_THRESHOLD = 0.8

// A Domain Service — a plain class with no framework decorators (it's not registered in the
// NestJS DI container either. The Application layer creates it directly with `new` when needed).
//
// The judgment "the original payment must be COMPLETED, and the refund amount can't exceed
// the payment amount" can't be made by Payment alone or Refund alone. Payment doesn't know
// about refund attempts against itself (a refund exists only as a Refund Aggregate), and
// Refund doesn't know the original payment's amount·status (it only references it via
// paymentId). Making this judgment requires loading both Aggregates and comparing them side
// by side, so this coordination logic can't go on either Aggregate's method (doing so would
// require receiving the entire other Aggregate as a parameter, collapsing the boundary) — it
// belongs here, in a separate Domain Service. (See the root docs/architecture/domain-service.md.)
export class RefundEligibilityService {
  public evaluate(
    payment: Payment,
    refund: Refund,
    mlFraudRiskScore: number
  ): RefundDecision {
    if (payment.status !== PaymentStatus.COMPLETED) {
      return { approved: false, reason: PaymentErrorMessage['A refund can only be requested for a completed payment.'] }
    }
    if (refund.amount > payment.amount) {
      return { approved: false, reason: PaymentErrorMessage['The refund amount cannot exceed the payment amount.'] }
    }
    if (mlFraudRiskScore >= ML_FRAUD_RISK_REJECTION_THRESHOLD) {
      return { approved: false, reason: PaymentErrorMessage['This refund pattern was flagged as high risk by the fraud-risk model and requires manual review.'] }
    }
    return { approved: true }
  }
}
